torch/backends/_nnapi/prepare.py

```diff
@@ -0,0 +1,167 @@
+        outs = [ torch.empty_like(out) for out in self.out_templates ]
+        
+        assert len(args) == len(self.inp_mem_fmts)
+        fixed_args = []
+        for idx in range(len(args)):
+            fmt = self.inp_mem_fmts[idx]
+            if fmt == 0:
+                fixed_args.append(args[idx].contiguous())
+            elif fmt == 1:
+                fixed_args.append(args[idx].permute(0,2,3,1).contiguous())
+            else:
+                raise Exception("Invalid mem_fmt")
+        comp.run(fixed_args, outs)
+        assert len(outs) == len(self.out_mem_fmts)
+        for idx in range(len(self.out_templates)):
+            fmt = self.out_mem_fmts[idx]
+            if fmt == 0:
+                pass
+            elif fmt == 1:
+                outs[idx] = outs[idx].permute(0,3,1,2)
+            else:
+                raise Exception("Invalid mem_fmt")
+        return outs
+    
+
+class NnapiInitWrapper(torch.nn.Module):
+    """Wrapper module to ensure NNAPI init is called."""
+    def __init__(self, nnapi_module):
+        super().__init__()
+        self.nnapi_module = nnapi_module
+        
+    def forward(self, args: List[torch.Tensor]) -> List[torch.Tensor]:
+        return self.nnapi_module(args)
+    
+    @torch.jit.export
+    def __getstate__(self):
+        return self.nnapi_module
+        
+    @torch.jit.export
+    def __setstate__(self, nnapi_module):
+        self.training = False
+        self.nnapi_module = nnapi_module
+        self.nnapi_module.init()
+
+
+class ListWrapper(torch.nn.Module):
+    """NNAPI list-ifying wrapper.
+
+    NNAPI always expects a list of inputs.  This module provides a
+    single-tensor input interface for models that want it.
+    """
+    def __init__(self, mod):
+        super().__init__()
+        self.mod = mod
+        
+    def forward(self, t: torch.Tensor) -> List[torch.Tensor]:
+        return self.mod([t])
+        
+class DelistWrapper(torch.nn.Module):
+    """NNAPI de-list-ifying wrapper.
+
+    NNAPI always provides a list of outputs.  This module provides a
+    single-tensor output interface for models that want it.
+    """
+    def __init__(self, mod):
+        super().__init__()
+        self.mod = mod
+        
+    def forward(self, ts: List[torch.Tensor]) -> torch.Tensor:
+        outs = self.mod(ts)
+        assert len(outs) == 1
+        return outs[0]
+
+class ListDelistWrapper(torch.nn.Module):
+    """NNAPI list-ifying and de-list-ifying wrapper.
+
+    NNAPI always expects a list of inputs and provides a list of outputs.
+    This module provides a single-tensor input/output interface
+    for models that want it.
+    """
+    def __init__(self, mod):
+        super().__init__()
+        self.mod = mod
+        
+    def forward(self, t: torch.Tensor) -> torch.Tensor:
+        outs = self.mod([t])
+        assert len(outs) == 1
+        return outs[0]
+
+
+def convert_model_to_nnapi(model, inputs):
+    model = torch.jit.freeze(model)
```

Comment:
(for my reference) Okay - it seems like freeze is a way to inline things. https://pytorch.org/docs/stable/generated/torch.jit.freeze.html

Reply:
Yes, this eliminates a *lot* of TorchScript cruft so we don't have to understand it during conversion.